src/main/java/fi/aalto/cs/apluscourses/ui/exercise/ExercisesView.java

```diff
@@ -1,26 +1,53 @@
 package fi.aalto.cs.apluscourses.ui.exercise;
 
+import static fi.aalto.cs.apluscourses.utils.PluginResourceBundle.getText;
+
 import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.application.ModalityState;
 import fi.aalto.cs.apluscourses.intellij.actions.ActionUtil;
 import fi.aalto.cs.apluscourses.intellij.actions.OpenSubmissionAction;
+import fi.aalto.cs.apluscourses.presentation.MainViewModel;
 import fi.aalto.cs.apluscourses.presentation.exercise.ExercisesTreeViewModel;
 import fi.aalto.cs.apluscourses.ui.GuiObject;
 import fi.aalto.cs.apluscourses.ui.base.TreeView;
+
+import java.awt.CardLayout;
+
+import javax.swing.JLabel;
 import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.SwingConstants;
+
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+
 public class ExercisesView {
   private TreeView exerciseGroupsTree;
+  private JLabel emptyText;
   private JPanel basePanel;
-
+  private JScrollPane pane;
   @GuiObject
   public JPanel toolbarContainer;
+  private JPanel cardPanel;
+  private CardLayout cl;
+  private MainViewModel mainViewModel;
```

Comment:
The fact that `ExercisesView` has a reference to the `MainViewModel` is a bit questionable IMO, which now makes me think that the "`ExercisesTreeViewModel` is never null" might have been a cleaner approach. Of course, I have the benefit of hindsight here, so I'm not saying that I knew this beforehand :smile: 

Anyways, I don't this needs to be changed at the moment. We can consider later if this is something that we want to modify.

Reply:
I agree it's a bit questionable and I agree it's an issue that doesn't have to be solved now.

Basically, how this could be solved, is putting `getCardToBeShown` in an interface (implemented by MainViewModel) and use that interface here.

Another option would be doing something like I suggested [here](https://github.com/Aalto-LeTech/intellij-plugin/pull/490#discussion_r564794544).

Reply:
Unless I misunderstood something, I still think it belongs in `ExercisesTreeViewModel` instead.

An interface that main view model implements would fix the broken encapsulation here, but I think the more correct solution would be to not pass main view model to the ctor at all.

Reply:
Yeah, you're right. Or better yet, there should probably be a class between `MainViewModel` and `ExercisesTreeViewModel` so the composition would be `[MainViewModel]<>1---1[ThatClass]<>0..1---1[ExercisesTreeViewModel]`.